contrib/polarity-for-haskellers/README.md

```diff
@@ -0,0 +1,115 @@
+2. instead of demanding the compontents when building the stream, ask for the decision of what we are going to observe
+
+The former is pretty simple, we define the datatype:
+```haskell 
+type StreamObservation :: SType -> SType -> SType
+data StreamObservation a r
+  = Hd (a -> r)
+  | Tl (Stream a -> r)
+```
+As you can see, everything is a strict datatype still so you know we are not cheating you. This datatype now describes what it means to observe a stream:
+1. either, you see the head (`Hd`) in which case you may do something with the element you find there, or
+2. you will observe the tail (`Tl`) in which case you may do something with the stream that makes up tail.
+
+Now, to restore the interface that we are used to, let's wrap the idea of "give us an observation and we will execute it for you" in a `newtype`:
+```haskell
+type Stream :: SType -> LType
+newtype Stream a = MkStream {unStream :: forall r. StreamObservation a r -> r}
+```
+As you may have noticed, this returns an `LType`, not an `SType`; the reason is that representationally, a `Stream` is now a function closure, the only primitive lazy datastructure in many strict langauges, just like in our imaginary strict Haskell.
+In polarity, which does not have functions but instead has `codata` types, we can write this type out a bit simpler, namely as:
+```polarity
+codata Stream(a : Type) 
+  { Stream(a).head(a : Type): a
+  , Stream(a).tail(a : Type): Stream(a)
+  }
+```
+Just like the Haskell version, it offers you two options to observe a stream, but it leaves out the CPS encoding part. This makes it also nicer to create streams with `codef`, which we will see in a minute. In order to do so in haskell, let's define the stream that consists only of ones, that is no matter how far you go (by observing `Tl`), every time you observe head afterwards (`Hd`), you obtain a one.
+```haskell
+ones :: Stream I
+ones = MkStream \case
+  Hd k -> k (I# 1#) -- observe the number one that lives in every cell of the Stream
+  Tl k -> k ones    -- observe the remainder of the stream
+ where
+  one = I# 1#
+```
+in polarity, the definition looks quite similar, except that, instead of using continuation passing style, we can write it in direct style, similarly to how it is done in object oriented programming's class methods:
+```polarity
+data Nat {Z : Nat, S(n : Nat) : Nat}
+
+codef Ones: Stream(Nat) { -- use codef to define a Stream
+  .head(_) => S(Z),       -- when observing the head, return 1
+  .tail(_) => Ones        -- otherwise, we observe a stream of ones
+}
+```
+
+So far so good, now we know how to use the non-dependent portion of polarity and how it corresponds to Haskell.
+
+Howerver, there's another peculiar thing happening: instead of infinitely many numbers, we now only need a single one to construct the stream, so, if we do not duplicate the stream or let the continuation use the result multiple times, there's actually only a single one we can get out of a stream.
```

Comment:
```suggestion
However, there's another peculiar thing happening: instead of infinitely many numbers, we now only need a single one to construct the stream, so, if we do not duplicate the stream or let the continuation use the result multiple times, there's actually only a single one we can get out of a stream.
```